File: src/daemon/fork.rs
use global_placeholders::global;
use std::{ffi::CString, process::exit};

#[allow(dead_code)]
pub enum Fork {
    Parent(libc::pid_t),
    Child,
}

pub fn chdir() -> Result<libc::c_int, i32> {
    let dir = CString::new(global!("pmc.base")).expect("CString::new failed");
    let res = unsafe { libc::chdir(dir.as_ptr()) };
    match res {
        -1 => Err(-1),
        res => Ok(res),
    }
}

pub fn fork() -> Result<Fork, i32> {
    let res = unsafe { libc::fork() };
    match res {
        -1 => Err(-1),
        0 => Ok(Fork::Child),
        res => Ok(Fork::Parent(res)),
    }
}

pub fn setsid() -> Result<libc::pid_t, i32> {
    let res = unsafe { libc::setsid() };
    match res {
        -1 => Err(-1),
        res => Ok(res),
    }
}

pub fn close_fd() -> Result<i32, i32> {
    let mut res = false;
    for i in 0..=2 {
        res |= unsafe { libc::close(i) } == -1;
    }

    match res {
        true => Err(-1),
        false => Ok(1),
    }
}

pub fn daemon(nochdir: bool, noclose: bool) -> Result<Fork, i32> {
    match fork() {
        Ok(Fork::Parent(_)) => exit(0),
        Ok(Fork::Child) => setsid().and_then(|_| {
            if !nochdir {
                chdir()?;
            }
            if !noclose {
                close_fd()?;
            }
            fork()
        }),
        Err(n) => Err(n),
    }
}

File: lib/fork.cc
#include "include/fork.h"

#include <cstdlib>
#include <cstring>
#include <iostream>
#include <stdexcept>
#include <unistd.h>

#ifdef _WIN32
#include <windows.h>
#else
#include <pwd.h>
#include <unistd.h>
#endif

using namespace std;

std::string home() {
#ifdef _WIN32
  const char *userProfile = std::getenv("USERPROFILE");
  if (userProfile) {
    return std::string(userProfile);
  } else {
    return "";
  }
#else
  struct passwd *pw = getpwuid(getuid());
  if (pw && pw->pw_dir) {
    return std::string(pw->pw_dir);
  } else {
    return "";
  }
#endif
}

Fork fork_process() {
  pid_t res = ::fork();
  if (res == -1) {
    perror("fork_process");
    throw std::runtime_error("fork() failed");
  } else if (res == 0) {
    return Fork::Child;
  } else {
    return Fork::Parent;
  }
}

pid_t set_sid() {
  pid_t res = ::setsid();
  if (res == -1) {
    perror("set_sid");
    throw std::runtime_error("setsid() failed");
  }
  return res;
}

void close_fd() {
  bool res = false;
  for (int i = 0; i <= 2; ++i) {
    res |= (::close(i) == -1);
  }

  if (res) {
    perror("close_fd");
    throw std::runtime_error("close_fd() failed");
  }
}

int32_t try_fork(bool nochdir, bool noclose, Callback callback) {
  try {
    Fork forkResult = fork_process();
    if (forkResult == Fork::Parent) {
      exit(0);
    } else if (forkResult == Fork::Child) {
      set_sid();
      if (!nochdir) {
        std::string home_dir = home() + ".pmc";
        chdir(home_dir.c_str());
      }
      if (!noclose) {
        close_fd();
      }
      forkResult = fork_process();
    }
    return static_cast<int32_t>(forkResult);
  } catch (const std::exception &e) {
    std::cerr << "[PMC] (cc) Error setting up daemon handler\n";
    perror("try_fork");
  }

  callback();
  return -1;
}